### PR TITLE
perf: replace pure-Python temperature sampling with a Rust implementation

### PR DESCRIPTION
**Pull request by ericcurtin**
## What
`vllm_vulkan/server.py` (the standalone Rust-`VulkanModel`-backed serving path this crate documents as giving "~3 tok/s on GB10") called `model.forward()` then passed its `Vec<f32>` return value — converted by PyO3 into 262144 individual Python `float` objects for Gemma4-E2B's vocab — into a pure-Python `temperature_sample()` that did a full `sorted()` over all of them (plus several more full-vocab list comprehensions for temperature scaling and softmax), **on every single decode step**.

## Measured impact
On this hardware, at the real vocab size (262144):
- Old pure-Python `temperature_sample`: **~82.2ms/call**
- New, through the standalone `sample_logits` pyfunction (which still pays a Python-list round trip): **~8.6ms/call — a 9.5x speedup**
- Pure Rust algorithmic cost alone (no Python interpreter involved, what `VulkanModel.forward_and_sample` actually uses): **~6.3ms/call** (`examples/bench_sampling.rs`)

At the documented "~3 tok/s" (≈333ms/token), the old sampling step alone was consuming roughly a quarter of total per-token latency.

## What changed
- **`model::sample_with_temperature`/`argmax`** (`src/model.rs`): the sampling algorithm itself, mirroring `temperature_sample`'s math — a single combined descending-probability sort instead of Python's two separate ones (it already serves both the top-k selection and the top-p walk). `uniform_random` is caller-supplied (e.g. Python's `random.random()`) rather than generated internally, so this crate doesn't need a `rand` dependency.
- **`VulkanModel.forward_and_sample`** (`src/lib.rs`): runs a decode step and samples the next token without the logit vector ever leaving Rust — replaces `forward()` + `greedy_sample()`/`temperature_sample()` in `generate()`.
- **`sample_logits`** (`src/lib.rs`): standalone pyfunction wrapping just the sampling step, for direct Python callers/testing.
- **`vllm_vulkan/server.py`**: `generate()` now calls `model.forward_and_sample(...)`; old `greedy_sample`/`temperature_sample` are kept (docstrings updated) for any external callers already depending on their exact signatures.

## Testing
- `model::sampling_tests` (6 new tests): `argmax` correctness, the `temperature < 0.01` → greedy threshold (matches `generate()`'s original dispatch logic — see doc comments for why this differs from `temperature_sample`'s own narrower `temperature == 0.0` check), `top_k=1` always returning the single candidate, `uniform_random=0.0` boundary behavior, and a 20-trial × 4-parameter-set × 5-uniform-random cross-check against a literal, unoptimized port of the original Python algorithm (including its two separate sorts).
- `examples/bench_sampling.rs`: pure-Rust algorithmic cost benchmark.
- `ruff check`/`ruff format --check`/`mypy` clean on `vllm_vulkan/server.py`.
- Manual smoke test from Python (built via `maturin develop --release`): confirmed greedy/small-temperature/top_k=1 behavior matches expectations.
- All 19 lib tests pass (`cargo test --release`), zero new clippy warnings (48 lib / 49 lib+all-targets, unchanged from `main`).

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -83,6 +83,23 @@ fn get_memory_info(device_idx: usize) -> PyResult<(u64, u64)> {
...
+/// for why that matters. `uniform_random` should be a fresh uniform `[0, 1)`
+/// draw per call (e.g. Python's `random.random()`); `top_k <= 0` means no
+/// top-k filtering (use the full vocab); `temperature < 0.01` means greedy
+/// (argmax) sampling, ignoring `top_p`/`top_k`/`uniform_random` (see
+/// `model::sample_with_temperature`'s doc comment for why 0.01 specifically).
+#[pyfunction]
+#[pyo3(signature = (logits, temperature=1.0, top_p=1.0, top_k=64, uniform_random=0.0))]
+fn sample_logits(logits: Vec<f32>, temperature: f32, top_p: f32, top_k: i64, uniform_random: f32) -> usize {
+    model::sample_with_temperature(&logits, temperature, top_p, top_k, uniform_random)
+}
```
[the remark this answers is not in the document]
Comment (ericcurtin) [reply src/lib.rs:111]: Fixed in 4292bbb — sample_logits now returns PyResult<usize> and raises a PyRuntimeError (matching this codebase's existing error-handling convention, e.g. get_device_info) instead of panicking on an empty logits vector. Verified from Python: raises a catchable RuntimeError, non-empty input still works correctly.
